### DISCUSSION TITLE
Does Grizzly remove dashboards from Grafana when removed from code?

### DISCUSSION THREAD

**Discussion by mikaello** [grafana/grizzly#490] [Q&A] [answered]
Is it possible to use Grizzly to sync removals of dashboards?

I mean, if I got a `main.jsonnet` like this which I have `grr apply`ed:

```jsonnet
{
  dashboards: [
    grr.dashboard.new('example', (import 'dashboards/alerts.jsonnet')),
    grr.dashboard.new('slo', (import 'dashboards/slo.jsonnet')),
  ],
}
```

```
$ grr apply main.jsonnet
Applying 2 resources
Dashboard.example added
Dashboard.slo added
2 resources added
```

and I later remove `slo`:

```jsonnet
{
  dashboards: [
    grr.dashboard.new('example', (import 'dashboards/alerts.jsonnet')),
  ],
}
```

... and apply this, the `slo` dashboard is not removed:

```
$ grr apply main.jsonnet
Applying 1 resources
Dashboard.example unchanged
1 resource unchanged
```

How can I instrument Grizzly to sync removals of dashboards? If Grizzly is not capable of doing this, what can I do to fix this?

**Comment by malcolmholmes** [comment, marked as answer]
No, Grizzly doesn't support this. Grizzly, in this case, was modelled upon Grafana Tanka. Tanka handles this with its "purge" functionality, where it stores an annotation against resources so they can be tracked back to originating in Tanka. Unfortunately, Grafana doesn't have any support for metadata such as these annotations, making it much harder to implement. How would we know if a dashboard (or other resource) was created manually, or has been removed from your jsonnet (or other) code?

**Reply by mikaello** [reply]
I see your point, thanks for your explanation.

> How would we know if a dashboard (or other resource) was created manually, or has been removed from your jsonnet (or other) code?

It must have been if Grizzly considered the code the authoritative source of truth, but I understand that this is not the case.

**Reply by malcolmholmes** [reply]
> It must have been if Grizzly considered the code the authoritative source of truth, but I understand that this is not the case

The problem is that when an item is removed from code, it is no-longer available to Grizzly in its source form. So either Grizzly would need to maintain state (as Terraform does), which would introduce a significant level of complexity, or remote objects would need to be annotated (as Tanka does), but we cannot do this without Grafana supporting annotations.

Find a way to recognise a remote resource as Grizzly provisioned, then we can implement purging.

**Comment by kshpytsya** [comment]
Related: #557
